doc/prebuilt/man/cyclictest.1
'\" t
.\"     Title: cyclictest
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 2008/04/01
.\"    Manual: Xenomai Manual
.\"    Source: Xenomai 2.99.5
.\"  Language: English
.\"
.TH "CYCLICTEST" "1" "2008/04/01" "Xenomai 2\&.99\&.5" "Xenomai Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
cyclictest \- Xenomai high resolution timer test
.SH "SYNOPSIS"
.sp
\fBcyclictest\fR [options]
.SH "DESCRIPTION"
.sp
\fBcyclictest\fR is part of the Xenomai test suite and tests the POSIX skin of Xenomai with a cyclic timer test\&.
.sp
For this program to work, you need to run a suitable Xenomai enabled kernel with the respective module (xeno_posix)\&.
.SH "OPTIONS"
.sp
\fBcyclictest\fR accepts the following options:
.PP
\fB\-b USEC, \-\-breaktrace=USEC\fR
.RS 4
send break trace command when latency > USEC
.RE
.PP
\fB\-c CLOCK, \-\-clock=CLOCK\fR
.RS 4
select clock: 0 = CLOCK_MONOTONIC (default) 1 = CLOCK_REALTIME
.RE
.PP
\fB\-d DIST, \-\-distance=DIST\fR
.RS 4
distance of thread intervals in us default=500
.RE
.PP
\fB\-i INTV, \-\-interval=INTV\fR
.RS 4
base interval of thread in us default=1000
.RE
.PP
\fB\-l LOOPS, \-\-loops=LOOPS\fR
.RS 4
number of loops: default=0 (endless)
.RE
.PP
\fB\-n, \-\-nanosleep\fR
.RS 4
use clock_nanosleep
.RE
.PP
\fB\-p PRIO, \-\-prio=PRIO\fR
.RS 4
priority of highest prio thread
.RE
.PP
\fB\-q, \-\-quiet\fR
.RS 4
print only a summary on exit
.RE
.PP
\fB\-r, \-\-relative\fR
.RS 4
use relative timer instead of absolute
.RE
.PP
\fB\-t NUM, \-\-threads=NUM\fR
.RS 4
number of threads: default=1
.RE
.PP
\fB\-v, \-\-verbose\fR
.RS 4
output values on stdout for statistics

format: n:c:v n=tasknum c=count v=value in us
.RE
.SH "AUTHOR"
.sp
\fBcyclictest\fR was written by Thomas Gleixner\&. This man page was written by Roland Stigge\&.


ChangeLog
2009-03-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vxworks/semLib.c (sem_destroy_internal) : Add missing
	rescheduling. Fix sem_read_proc().

2009-03-19  Philippe Gerum  <rpm@xenomai.org>

	* include/asm-generic/wrappers.h, src/nucleus/shadow.c: Fix
	generic bits for 2.6.29.

2009-03-18  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/pipe.c: Fix streaming mode.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/nucleus/shadow.c: Trace syscall exits.

	* patching file include/asm-generic/syscall.h,
	ksrc/nucleus/shadow.c: Add xn_sys_current_info.

	* src/rtdk/rt_print.c: suspend/resume printer thread.

	* ksrc/skins/native/syscall.c, src/skins/native/task.c: Support
	T_WARNSW on task creation.

	* src/skins/native/task.c: Avoid lazy non-TLS rt_task_self.

	* src/skins/native/task.c: Silent migration before self-deletion.

	* src/skins, ksrc/nucleus/shadow.c: Rework initial Linux
	schedparam setting.

	* ksrc/nucleus, include/asm-generic, ksrc/arch/generic: Exclude
	CPUs from Xenomai use.

	* ksrc/nucleus/shadow.c: Cleanup taskexit hook.

	* ksrc/nucleus/shadow.c: Suppress switch warning upon signal
	receipt when the process is being debugged.

	* ksrc/nucleus/sched.c: Allow to change the watchdog timeout via a
	module parameter.

	* ksrc/arch/x86/usercopy_{32, 64}.c: Fix miscompilation of
	__do_strncpy_from_user.

	* src/rtdk, include/rtdk.h: Add assert_context helpers.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/skins/posix/syscall.c: Do not return -EFAULT on zero-length
	names. Instead, return -EINVAL when trying to create objects with
	empty names.

2009-03-13  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/mx.c: Streamline mutex support.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/task.c, src/skins/vrtx/task.c: Fix task creation.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/shadow.c, ksrc/nucleus/synch.c: Introduce debug
	option to detect mutexes held in relaxed sections.

2009-03-07  Philippe Gerum  <rpm@xenomai.org>

	* src/skins/posix/shm.c (__wrap_munmap): Forward to glibc upon EBADF.

	* ksrc/nucleus/Config.in, ksrc/arch/{powerpc, x86}/Config.in: Add
	missing SYS_STACKPOOLSZ switch.

2009-02-26  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/powerpc/patches: Update Adeos support to
	2.6.24.6-powerpc-DENX-2.4-03 (compatible with MPC5121-ADS
	hardware, see ksrc/arch/powerpc/patches/README).

2009-02-25  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/nmi.c: Rework NMI watchdog support.
	
2008-02-24  Stephane Fillod  <stephane.fillod@thomson.net>

	* src/skins/posix/shm.c (__wrap_mmap64): Make real call depend on
	configure detection (HAVE_MMAP64).
	
2009-02-24  Yi Li <liyi.dev@gmail.com>

	* configure.in: Remove -mfdpic switch. Set by default by
	bfin-linux-uclibc-gcc, and may cause error with bfin-uclinux-gcc.

2009-02-24  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/{arm, blackfin, powerpc}/hal.c (rthal_timer_request):
	Port x86 fix of race window.

2009-02-24  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/hal-common.c (rthal_timer_request): Fix race.

	* ksrc/nucleus/shadow.c (gatekeeper_thread): Do not refer to
	uninit sched slots.

2009-02-23  Paul Corner <paul_c@tuxcnc.org>

	* debian/prepare-patch.sh: Fix patch generation rules.

2009-02-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.19-x86-2.2-06,
	2.6.28.7-x86-2.2-06.

2009-02-17  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-05.

	* ksrc/arch/powerpc/patches: Upgrade to
	2.6.28.5-powerpc-DENX-2.4-04.

2009-02-13  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* src/skins/posix/shm.c (__wrap_mmap, __wrap_mmap64): Map only
	part of the underlying nucleus heap.

	* ksrc/nucleus/heap.c (xnheap_mmap): Allow partial mapping in
	user-space of a nucleus heap.

	* ksrc/nucleus/sched.c, include/nucleus/sched.h: Introduce
	sched->cpu field member.

2009-02-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pod.c (xnpod_start_thread): Run start hook for
	user-space threads as well.

2009-02-11  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/posix/shm.c (ftruncate): Remove GFP_USER hint.

2009-02-07  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/blackfin, include/asm-blackfin: Prepare for 2.6.28.

	* ksrc/arch/blackfin/patches: Upgrade to 2.6.28-1.9-00 (GIT).

	* ksrc/nucleus, include/nucleus: Introduce host tick deferral.

2009-02-02  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/syscall.c (rt_queue_write): Fix memory leak.

2009-02-01  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/heap, include/asm-generic/system.h, ksrc/skins: Fix
	auto-cleanup for shared heaps.

2009-01-31  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/drivers/testing/switchtest.c,
	  src/testsuite/switchtest/switchtest.c: Add stress mode to
	switchtest.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/ia64, include/asm-ia64: Itanium port discontinued.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-04,
	2.6.28.2-x86-2.2-05.

2009-01-27  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* include/asm-x86/bits/pod_{32, 64}.h (xnarch_leave_root): Fix FPU
	backup clobbering.
	
2009-01-27  Philippe Gerum  <rpm@eclipse.xenomai.org>

	* ksrc/nucleus, include/nucleus, ksrc/skins/posix, include/posix,
	sim, include/asm-sim: Introduce POSIX sporadic scheduling (PSS)
	class.

2009-01-21  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/arch/arm/patches: Upgrade to 2.6.28-arm-1.2-00.
	
2009-01-19  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pipe.c, ksrc/skins/native/pipe.c,
	ksrc/skins/rtai/fifo.c: Refactor callback management, fixing a
	severe latency issue, and a potential memory leak in both
	disconnect/release handlers.

	* ksrc/nucleus/pipe.c (xnpipe_disconnect, xnpipe_release):

testsuite/cyclic/cyclictest.c
/*
 * High resolution timer test software
 *
 * Copyright (C) 2005,2006 Thomas Gleixner <tglx@linutronix.de>
 * (Enhanced by the Xenomai crew)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License Version
 * 2 as published by the Free Software Foundation;
 *
 */

#define VERSION_STRING "V 0.11xn"

#define INGO_TRACE 0
#define TGLX_TRACE 0
#ifndef IPIPE_TRACE
#define IPIPE_TRACE 0
#endif

#include <fcntl.h>
#include <getopt.h>
#include <pthread.h>
#include <signal.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <limits.h>

#include <linux/unistd.h>

#include <sys/stat.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/mman.h>
#include <cobalt/trace.h>

#define ARRAY_SIZE(x) (sizeof(x) / sizeof((x)[0]))

/* Ugly, but .... */
#define gettid() syscall(__NR_gettid)
#define sigev_notify_thread_id _sigev_un._tid

extern int clock_nanosleep (clockid_t __clock_id, int __flags,
			    __const struct timespec *__req,
			    struct timespec *__rem);

#define USEC_PER_SEC	1000000
#define NSEC_PER_SEC	1000000000

#define MODE_CYCLIC		0
#define MODE_CLOCK_NANOSLEEP	1
#define MODE_SYS_ITIMER		2
#define MODE_SYS_NANOSLEEP	3
#define MODE_SYS_OFFSET		2

#define TIMER_RELTIME	0

/* Must be power of 2 ! */
#define VALBUF_SIZE	16384

/* Struct to transfer parameters to the thread */
struct thread_param {
	int prio;
	int mode;
	int timermode;
	int signal;
	int clock;
	unsigned long max_cycles;
	struct thread_stat *stats;
	int bufmsk;
	unsigned long interval;
};

/* Struct for statistics */
struct thread_stat {
	unsigned long cycles;
	unsigned long cyclesread;
	long min;
	long max;
	long act;
	double avg;
	long *values;
	pthread_t thread;
	int threadstarted;
	int tid;
	int traced;
};

static int test_shutdown;
static int tracelimit = 100000;
static struct timespec start;

static inline void tsnorm(struct timespec *ts)
{
	while (ts->tv_nsec >= NSEC_PER_SEC) {
		ts->tv_nsec -= NSEC_PER_SEC;
		ts->tv_sec++;
	}
}

static inline long calcdiff(struct timespec t1, struct timespec t2)
{
	long diff;
	diff = USEC_PER_SEC * ((int) t1.tv_sec - (int) t2.tv_sec);
	diff += ((int) t1.tv_nsec - (int) t2.tv_nsec) / 1000;
	return diff;
}

/*
 * timer thread
 *
 * Modes:
 * - clock_nanosleep based
 * - cyclic timer based
 *
 * Clock:
 * - CLOCK_MONOTONIC
 * - CLOCK_REALTIME
 * - CLOCK_MONOTONIC_HR
 * - CLOCK_REALTIME_HR
 *
 */
static void *timerthread(void *param)
{
	struct thread_param *par = param;
	struct sched_param schedp;
	sigset_t sigset;
	struct timespec now, next, interval;
	struct thread_stat *stat = par->stats;
	int policy = par->prio ? SCHED_FIFO : SCHED_OTHER;
	int err;
#ifdef __UNSUPPORTED
	struct itimerval itimer;
	struct sigevent sigev;
	timer_t timer;
	struct itimerspec tspec;
#endif
#if (INGO_TRACE + TGLX_TRACE)
	int stopped = 0;
#endif

	interval.tv_sec = par->interval / USEC_PER_SEC;
	interval.tv_nsec = (par->interval % USEC_PER_SEC) * 1000;

#if INGO_TRACE
	system("echo 1 > /proc/sys/kernel/trace_all_cpus");
	system("echo 1 > /proc/sys/kernel/trace_enabled");
	system("echo 1 > /proc/sys/kernel/trace_freerunning");
	system("echo 0 > /proc/sys/kernel/trace_print_at_crash");
	system("echo 1 > /proc/sys/kernel/trace_user_triggered");
	system("echo 0 > /proc/sys/kernel/trace_user_trigger_irq");
	system("echo 0 > /proc/sys/kernel/trace_verbose");
	system("echo 0 > /proc/sys/kernel/preempt_thresh");
	system("echo 0 > /proc/sys/kernel/wakeup_timing");
	system("echo 0 > /proc/sys/kernel/preempt_max_latency");
#endif

	stat->tid = gettid();

	sigemptyset(&sigset);
	sigaddset(&sigset, par->signal);
	sigprocmask(SIG_BLOCK, &sigset, NULL);

#ifdef __UNSUPPORTED
	if (par->mode == MODE_CYCLIC) {
		sigev.sigev_notify = SIGEV_THREAD_ID | SIGEV_SIGNAL;
		sigev.sigev_signo = par->signal;
		sigev.sigev_notify_thread_id = stat->tid;
		timer_create(par->clock, &sigev, &timer);
		tspec.it_interval = interval;
	}
#endif

	memset(&schedp, 0, sizeof(schedp));
	schedp.sched_priority = par->prio;
	err = pthread_setschedparam(pthread_self(), policy, &schedp);
#ifdef __XENO__
	if (err) {
	    fprintf(stderr, "pthread_setschedparam: %s\n"
		    "(modprobe xeno_posix?)\n", strerror(err));
	    test_shutdown = 1;
	    return (void *) 1;
	}
#endif

	/* Get current time */
	next = start;
	next.tv_sec++;

#ifdef __UNSUPPORTED
	if (par->mode == MODE_CYCLIC) {
		if (par->timermode == TIMER_ABSTIME)
			tspec.it_value = next;
		else {
			tspec.it_value.tv_nsec = 0;
			tspec.it_value.tv_sec = 1;
		}
		timer_settime(timer, par->timermode, &tspec, NULL);
	}

	if (par->mode == MODE_SYS_ITIMER) {
		itimer.it_value.tv_sec = 1;
		itimer.it_value.tv_usec = 0;
		itimer.it_interval.tv_sec = interval.tv_sec;
		itimer.it_interval.tv_usec = interval.tv_nsec / 1000;
		setitimer (ITIMER_REAL,  &itimer, NULL);
	}
#endif

	stat->threadstarted++;

#if (INGO_TRACE + TGLX_TRACE)
	gettimeofday(0,(struct timezone *)1);
#endif

	while (!test_shutdown) {

		long diff;
#ifdef __UNSUPPORTED
		int sigs;
#endif

		/* Wait for next period */
		switch (par->mode) {
#ifdef __UNSUPPORTED
		case MODE_CYCLIC:
		case MODE_SYS_ITIMER:
			if (__STD(sigwait(&sigset, &sigs)) < 0)
				goto out;
			break;
#endif

		case MODE_CLOCK_NANOSLEEP:
			if (par->timermode == TIMER_ABSTIME)
				clock_nanosleep(par->clock, TIMER_ABSTIME, &next, NULL);
			else {
				clock_gettime(par->clock, &now);
				clock_nanosleep(par->clock, TIMER_RELTIME, &interval, NULL);
				next.tv_sec = now.tv_sec + interval.tv_sec;
				next.tv_nsec = now.tv_nsec + interval.tv_nsec;
				tsnorm(&next);
			}
			break;

#ifdef __UNSUPPORTED
		case MODE_SYS_NANOSLEEP:
			clock_gettime(par->clock, &now);
			nanosleep(&interval, NULL);
			next.tv_sec = now.tv_sec + interval.tv_sec;
			next.tv_nsec = now.tv_nsec + interval.tv_nsec;
			tsnorm(&next);
			break;
#endif
		}
		clock_gettime(par->clock, &now);

		diff = calcdiff(now, next);
		if (diff < stat->min)
			stat->min = diff;
		if (diff > stat->max) {
			stat->max = diff;
#if IPIPE_TRACE
			if (stat->traced)
				xntrace_user_freeze(diff, 0);
#endif
		}
		stat->avg += (double) diff;

#if (INGO_TRACE + TGLX_TRACE)
		if (!stopped && (diff > tracelimit)) {
			stopped++;
			gettimeofday(0,0);
			test_shutdown++;
		}
#endif
		stat->act = diff;
		stat->cycles++;

		if (par->bufmsk)
			stat->values[stat->cycles & par->bufmsk] = diff;

		next.tv_sec += interval.tv_sec;
		next.tv_nsec += interval.tv_nsec;
		tsnorm(&next);

		if (par->max_cycles && par->max_cycles == stat->cycles)
			break;
	}

#ifdef __UNSUPPORTED
out:
	if (par->mode == MODE_CYCLIC)
		timer_delete(timer);

	if (par->mode == MODE_SYS_ITIMER) {
		itimer.it_value.tv_sec = 0;
		itimer.it_value.tv_usec = 0;
		itimer.it_interval.tv_sec = 0;
		itimer.it_interval.tv_usec = 0;
		setitimer (ITIMER_REAL,  &itimer, NULL);
	}
#endif

	/* switch to normal */
	schedp.sched_priority = 0;
	pthread_setschedparam(pthread_self(), SCHED_OTHER, &schedp);

	stat->threadstarted = -1;

	return NULL;
}


/* Print usage information */
static void display_help(void)
{
	printf("cyclictest %s\n", VERSION_STRING);
	printf("Usage:\n"
	       "cyclictest <options>\n\n"
	       "-b USEC  --breaktrace=USEC send break trace command when latency > USEC\n"
	       "-c CLOCK --clock=CLOCK     select clock\n"
	       "                           0 = CLOCK_MONOTONIC (default)\n"
	       "                           1 = CLOCK_REALTIME\n"
	       "-d DIST  --distance=DIST   distance of thread intervals in us default=500\n"
	       "-i INTV  --interval=INTV   base interval of thread in us default=1000\n"
	       "-l LOOPS --loops=LOOPS     number of loops: default=0(endless)\n"
	       "-n       --nanosleep       use clock_nanosleep\n"
	       "-p PRIO  --prio=PRIO       priority of highest prio thread\n"
	       "-q       --quiet	   print only a summary on exit\n"
	       "-r       --relative        use relative timer instead of absolute\n"
#ifdef __UNSUPPORTED
	       "-s       --system          use sys_nanosleep and sys_setitimer\n"
#endif
	       "-t NUM   --threads=NUM     number of threads: default=1\n"
	       "-v       --verbose         output values on stdout for statistics\n"
	       "                           format: n:c:v n=tasknum c=count v=value in us\n");
	exit(0);
}

static int use_nanosleep = MODE_CLOCK_NANOSLEEP; /* make this default for now */
static int timermode  = TIMER_ABSTIME;
static int use_system;
static int priority = 99;
static int num_threads = 1;
static int max_cycles;
static int clocksel = 0;
static int verbose;
static int quiet;
static int interval;
static int distance;

static int clocksources[] = {
	CLOCK_MONOTONIC,
	CLOCK_REALTIME,
};

/* Process commandline options */
static void process_options (int argc, char *argv[])
{
	int error = 0;
	for (;;) {
		int option_index = 0;
		/** Options for getopt */
		static struct option long_options[] = {
			{"breaktrace", required_argument, NULL, 'b'},
			{"clock", required_argument, NULL, 'c'},
			{"distance", required_argument, NULL, 'd'},
			{"interval", required_argument, NULL, 'i'},
			{"loops", required_argument, NULL, 'l'},
			{"nanosleep", no_argument, NULL, 'n'},
			{"priority", required_argument, NULL, 'p'},
			{"quiet", no_argument, NULL, 'q'},
			{"relative", no_argument, NULL, 'r'},
#ifdef __UNSUPPORTED
			{"system", no_argument, NULL, 's'},
#endif
			{"threads", required_argument, NULL, 't'},
			{"verbose", no_argument, NULL, 'v'},
			{"help", no_argument, NULL, '?'},
			{NULL, 0, NULL, 0}
		};
		int c = getopt_long (argc, argv, "b:c:d:i:l:np:qrt:v",
			long_options, &option_index);
		if (c == -1)
			break;
		switch (c) {
		case 'b': tracelimit = atoi(optarg); break;
		case 'c': clocksel = atoi(optarg); break;
		case 'd': distance = atoi(optarg); break;
		case 'i': interval = atoi(optarg); break;
		case 'l': max_cycles = atoi(optarg); break;
		case 'n': use_nanosleep = MODE_CLOCK_NANOSLEEP; break;
		case 'p': priority = atoi(optarg); break;
		case 'q': quiet = 1; break;
		case 'r': timermode = TIMER_RELTIME; break;
#ifdef __UNSUPPORTED
		case 's': use_system = MODE_SYS_OFFSET; break;
#endif
		case 't': num_threads = atoi(optarg); break;
		case 'v': verbose = 1; break;
		case '?': error = 1; break;
		}
	}

	if (clocksel < 0 || clocksel > ARRAY_SIZE(clocksources))
		error = 1;

	if (priority < 0 || priority > 99)
		error = 1;

	if (num_threads < 1)
		error = 1;

	if (error)
		display_help ();
}

static void sighand(int sig)
{
	test_shutdown = 1;
}

static void print_stat(struct thread_param *par, int index, int verbose)
{
	struct thread_stat *stat = par->stats;

	if (!verbose) {
		if (!quiet)
			printf("T:%2d (%5d) P:%2d I:%8ld C:%8lu "
			       "Min:%8ld Act:%8ld Avg:%8ld Max:%8ld\n",
			       index, stat->tid, par->prio, par->interval,
			       stat->cycles, stat->min, stat->act,
			       stat->cycles ?
			       (long)(stat->avg/stat->cycles) : 0, stat->max);
	} else {
		while (stat->cycles != stat->cyclesread) {
			long diff = stat->values[stat->cyclesread & par->bufmsk];
			printf("%8d:%8lu:%8ld\n", index, stat->cyclesread, diff);
			stat->cyclesread++;
		}
	}
}

int main(int argc, char **argv)
{
	sigset_t sigset;
	int signum = SIGALRM;
	int mode;
	struct thread_param *par;
	struct thread_stat *stat;
	pthread_attr_t thattr;
	int i, ret = -1;

#ifndef __XENO__
	if (geteuid()) {
		printf("need to run as root!\n");
		exit(-1);
	}
#endif

	mlockall(MCL_CURRENT | MCL_FUTURE);

	if (CONFIG_XENO_DEFAULT_PERIOD > 1000000)
		interval = CONFIG_XENO_DEFAULT_PERIOD;
	else
		interval = 1000000;
	distance = interval / 2;

	process_options(argc, argv);

	mode = use_nanosleep + use_system;

	sigemptyset(&sigset);
	sigaddset(&sigset, signum);
	sigprocmask (SIG_BLOCK, &sigset, NULL);

	signal(SIGINT, sighand);
	signal(SIGTERM, sighand);

	par = calloc(num_threads, sizeof(struct thread_param));
	if (!par)
		goto out;
	stat = calloc(num_threads, sizeof(struct thread_stat));
	if (!stat)
		goto outpar;

	clock_gettime(clocksources[clocksel], &start);

	for (i = 0; i < num_threads; i++) {
		if (verbose) {
			stat[i].values = calloc(VALBUF_SIZE, sizeof(long));
			if (!stat[i].values)
				goto outall;
			par[i].bufmsk = VALBUF_SIZE - 1;
		}

		par[i].prio = priority;
		if (priority)
			priority--;
		par[i].clock = clocksources[clocksel];
		par[i].mode = mode;
		par[i].timermode = timermode;
		par[i].signal = signum;
		par[i].interval = interval;
		interval += distance;
		par[i].max_cycles = max_cycles;
		par[i].stats = &stat[i];
		stat[i].min = 1000000;
		stat[i].max = -1000000;
		stat[i].avg = 0.0;
		pthread_attr_init(&thattr);
		pthread_attr_setstacksize(&thattr, PTHREAD_STACK_MIN * 4);
		pthread_create(&stat[i].thread, &thattr, timerthread, &par[i]);
		stat[i].threadstarted = 1;
		stat[i].traced = (i == 0 && IPIPE_TRACE > 0);
	}

	while (!test_shutdown) {
		char lavg[256];
		int fd, len, allstopped;

		if (!verbose && !quiet) {
			fd = open("/proc/loadavg", O_RDONLY, 0666);
			len = read(fd, &lavg, 255);
			close(fd);
			lavg[len-1] = 0x0;
			printf("%s          \n\n", lavg);
		}

		allstopped = max_cycles ? 1 : 0;

		for (i = 0; i < num_threads; i++) {
			print_stat(&par[i], i, verbose);
			if (stat[i].cycles < max_cycles)
				allstopped = 0;
		}
		usleep(10000);
		if (test_shutdown || allstopped == num_threads)
			break;
		if (!verbose && !quiet)
			printf("\033[%dA", num_threads + 2);
	}
	if (quiet) {
		quiet = 0; /* Now we want to output the statistics */
		for (i = 0; i < num_threads; i++) {
			print_stat(&par[i], i, verbose);
		}
	}
	ret = 0;
 outall:
	test_shutdown = 1;
	for (i = 0; i < num_threads; i++) {
		if (stat[i].threadstarted > 0)
			pthread_kill(stat[i].thread, SIGTERM);
		if (stat[i].threadstarted)
			pthread_join(stat[i].thread, NULL);
		if (stat[i].values)
			free(stat[i].values);
	}
	free(stat);
 outpar:
	free(par);
 out:
	exit(ret);
}
